semgrep-core/matching/Rules_filter.ml

```diff
@@ -0,0 +1,155 @@
+    * => use Str.search_forward instead.
+    *)
+   try
+     Str.search_forward re str 0 |> ignore; true
+   with Not_found -> false
+  )
+
+let reserved_id lang str =
+  Metavars_generic.is_metavar_name str ||
+  (* in JS field names can be regexps *)
+  (lang = Lang.Javascript && Matching_generic.is_regexp_string str) ||
+  (* ugly hack that we then need to handle also here *)
+  str = AST_generic.special_multivardef_pattern ||
+  (* ugly: because ast_js_build introduce some extra "!default" ids *)
+  (lang = Lang.Javascript && str = Ast_js.default_entity) ||
+  (* parser_js.mly inserts some implicit this *)
+  (lang = Lang.Java && str = "this")
+
+let reserved_str str =
+  str = "..." ||
+  Matching_generic.is_regexp_string str
+
+(*****************************************************************************)
+(* ID extractor  *)
+(*****************************************************************************)
+let extract_specific_strings lang any =
+  let res = ref [] in
+  let visitor = V.mk_visitor {V.default_visitor with
+     V.kident = (fun (_k, _) (str, _tok) ->
+       if not (reserved_id lang str)
+       then Common.push str res
+     );
+     V.kexpr = (fun (k, _) x ->
+       (match x with
+       (* less: we could extract strings for the other literals too?
+        * atoms, chars, even int?
+        *)
+       | L (String (str, _tok)) ->
+         if not (reserved_str str)
+         then Common.push str res
+       (* do not recurse there, the type does not have to be in the source *)
+       | TypedMetavar _ ->
+          ()
+       | _ -> k x
+       );
+     );
+  } in
+  visitor any;
+  !res
+
+(*****************************************************************************)
+(* Entry point *)
+(*****************************************************************************)
+
+let filter_rules_relevant_to_file_using_regexp2 rules lang file =
+  let str = Common.read_file file in
+  rules |> List.filter (fun rule ->
+    let pat = rule.R.pattern in
+    let xs = extract_specific_strings lang pat in
+    (* pr2_gen xs; *)
+    let match_ =
+    (* we could avoid running multiple regexps on the same file
+     * by first orring them and do the and only of the or succeed,
+     * but probably not worth the opti.
+      let t = xs |> List.map (fun x -> regexp_matching_str x) |> Re.alt in
+      let re = compile_regexp t in
+      run_regexp re str
+      *)
+      (* Note that right now we do a for_all but it mighe be incorrect
+       * at some point if the pattern contains DisjExpr for example, in
+       * which case we will need extract_specific_strings to directly
+       * extract a complex regexp instead handling itself disjunction.
+       *)
+      xs |> List.for_all (fun x ->
+         let t = regexp_matching_str x in
+         let re = compile_regexp t in
+         run_regexp re str
+    )
+
+    in
+    if !Flag.debug && not match_
+    then pr2 (spf "filtering rule %s" rule.R.id);
```

Comment:
... and it's here 😛 . Can we add the filename to the debug message?

Reply:
It is already display before as PARSING foo.js

Reply:
👍